File: src/handlers/strava.ts
/**
 * Strava webhook handler.
 *
 * Strava webhooks have two phases:
 * 1. Validation: GET with hub.verify_token, hub.challenge, hub.mode
 * 2. Events: POST with JSON body containing activity/athlete event data
 *
 * Auth: Strava does NOT sign webhook payloads. We use a secret token in the
 * URL path as the only defense. The webhook callback URL should be registered as:
 *   https://webhooks.yourdomain.com/webhook/strava/<STRAVA_WEBHOOK_SECRET>
 */

import type { Config } from "../config.js";
import type { Logger } from "../logger.js";

export interface StravaEvent {
	object_type: string;
	object_id: number;
	aspect_type: string;
	owner_id: number;
	subscription_id: number;
	event_time: number;
	[key: string]: unknown;
}

export interface StravaHandlerResult {
	status: number;
	body?: unknown;
	payload?: StravaEvent;
}

export function validatePathSecret(
	pathSecret: string,
	config: Config,
	logger: Logger,
): boolean {
	if (!config.stravaWebhookSecret) {
		logger.warn(
			"STRAVA_WEBHOOK_SECRET not set, skipping path secret validation",
		);
		return true;
	}
	return pathSecret === config.stravaWebhookSecret;
}

export function handleStravaValidation(
	pathSecret: string,
	hubMode: string,
	hubChallenge: string,
	hubVerifyToken: string,
	config: Config,
	logger: Logger,
): StravaHandlerResult {
	if (!validatePathSecret(pathSecret, config, logger)) {
		logger.warn("Strava path secret mismatch on validation request");
		return { status: 404 };
	}

	if (hubMode !== "subscribe") {
		logger.warn("Unexpected hub.mode", { hub_mode: hubMode });
		return { status: 400 };
	}

	if (config.stravaVerifyToken && hubVerifyToken !== config.stravaVerifyToken) {
		logger.warn("Strava verify token mismatch");
		return { status: 403 };
	}

	logger.info("Strava webhook validation successful");
	return { status: 200, body: { "hub.challenge": hubChallenge } };
}

export function handleStravaWebhook(
	pathSecret: string,
	body: StravaEvent,
	config: Config,
	logger: Logger,
): StravaHandlerResult {
	if (!validatePathSecret(pathSecret, config, logger)) {
		logger.warn("Strava path secret mismatch on event delivery");
		return { status: 404 };
	}

	logger.info("Strava event received", {
		aspect_type: body.aspect_type,
		object_type: body.object_type,
		object_id: body.object_id,
		owner_id: body.owner_id,
	});

	return { status: 200, payload: body };
}

File: src/openclaw_api_server/handlers/strava.py
"""Strava webhook handler.

Strava webhooks have two phases:
1. Validation: GET with hub.verify_token, hub.challenge, hub.mode - must return hub.challenge
2. Events: POST with JSON body containing activity/athlete event data

Auth: Strava does NOT sign webhook payloads. We use a secret token in the URL path
as the only defense. The webhook callback URL should be registered as:
  https://webhooks.yourdomain.com/webhook/strava/<STRAVA_WEBHOOK_SECRET>
"""

import logging

from fastapi import APIRouter, Path, Query, Request, Response
from fastapi.responses import JSONResponse

from openclaw_api_server.config import config
from openclaw_api_server.gateway import forward_to_gateway

logger = logging.getLogger(__name__)
router = APIRouter()


def _validate_path_secret(path_secret: str) -> bool:
    if not config.strava_webhook_secret:
        logger.warning("STRAVA_WEBHOOK_SECRET not set, skipping path secret validation")
        return True
    return path_secret == config.strava_webhook_secret


@router.get("/webhook/strava/{path_secret}")
async def strava_validation(
    path_secret: str = Path(),
    hub_mode: str = Query(alias="hub.mode"),
    hub_challenge: str = Query(alias="hub.challenge"),
    hub_verify_token: str = Query(alias="hub.verify_token"),
) -> Response:
    """Handle Strava subscription validation callback."""
    if not _validate_path_secret(path_secret):
        return Response(status_code=404)

    if hub_mode != "subscribe":
        return Response(status_code=400)

    if config.strava_verify_token and hub_verify_token != config.strava_verify_token:
        logger.warning("Strava verify token mismatch")
        return Response(status_code=403)

    logger.info("Strava webhook validation successful")
    return JSONResponse({"hub.challenge": hub_challenge})


@router.post("/webhook/strava/{path_secret}")
async def strava_webhook(path_secret: str, request: Request) -> Response:
    """Handle Strava event notifications."""
    if not _validate_path_secret(path_secret):
        return Response(status_code=404)

    body = await request.json()

    object_type = body.get("object_type", "unknown")
    aspect_type = body.get("aspect_type", "unknown")
    object_id = body.get("object_id")
    owner_id = body.get("owner_id")

    logger.info("Strava event: %s %s (object=%s, owner=%s)", aspect_type, object_type, object_id, owner_id)

    await forward_to_gateway("strava", body)

    return Response(status_code=200)
